host/acquire_shard_integration_test.go

```diff
@@ -0,0 +1,175 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package host
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/suite"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+)
+
+// TestAcquireShard_OwnershipLostErrorSuite tests what happens when acquire shard returns an ownership lost error.
+func TestAcquireShard_OwnershipLostErrorSuite(t *testing.T) {
+	s := new(ownershipLostErrorSuite)
+	suite.Run(t, s)
+}
+
+// TestAcquireShard_DeadlineExceededErrorSuite tests what happens when acquire shard returns a deadline exceeded error
+func TestAcquireShard_DeadlineExceededErrorSuite(t *testing.T) {
+	s := new(deadlineExceededErrorSuite)
+	suite.Run(t, s)
+}
+
+// TestDoesNotRetry verifies that we do not retry acquiring the shard when we get an ownership lost error.
+func (s *ownershipLostErrorSuite) TestDoesNotRetry() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	numAttempts := 0
+	for {
+		select {
+		case msg := <-s.logs:
+			msg = strings.ToLower(msg)
+			if strings.Contains(msg, "error acquiring shard") {
+				numAttempts++
+				if strings.Contains(msg, "this error is retryable") {
+					s.Fail("should not be considered retryable")
+					return
+				}
+			}
+			if strings.Contains(msg, "acquired shard") {
+				s.Fail("should not acquire shard")
+				return
+			}
+			if strings.Contains(msg, "couldn't acquire shard") {
+				s.Equal(1, numAttempts, "should fail after one attempt")
+				return
+			}
+		case <-ctx.Done():
+			s.Fail("timed out waiting for shard update error")
+			return
+		}
+	}
+}
+
+// TestDoesRetry verifies that we do retry acquiring the shard when we get a deadline exceeded error.
+func (s *deadlineExceededErrorSuite) TestDoesRetry() {
```

Comment:
Can we change the failure rate to be less 1 so it could success after retry.

Reply:
I think the amount of non-determinism here would make this test unwieldy. We already have this case covered by a unit test.

Reply:
The major point of doing this integration test is to verify acquire shard can continue to work when there is retryable errors from persistence other than ShardOwnershipLost error.
I think the value of this integration test is greatly reduced if we cannot verify that.

Reply:
Done. I added a seed parameter to each faulty method to make this deterministic.